Added Evaluation Mode (#251)

mayankbucha opened the pull request:
Hi @Coteh, evaluation mode for the issue #251 is added. I have used `EVALUATION_MODE `flag, which when set to true adds evaluation mode and removes uploaded image after 5 mins. 
I have also converted `startDatabaseClient` function callbacks to promise, but they are simple promises and still looks like callbacks. Can we improve them?
Please review

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,64 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
             } else {
                 usersCollection = database.collection("users");
-                finishInitialization();
+                CheckForImageEntriesCollection(collectionNames);
```
```suggestion
                checkForImageEntriesCollection(collectionNames);
```

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,64 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
+                CheckForImageEntriesCollection(collectionNames);
             }
+        }).catch((err) => {
+            throw err;
```
```suggestion
            callback(err);
```

The `catch` below will not catch error if thrown here. `UnhandledPromiseRejectionWarning` will happen instead. Better to call the callback in this particular case.

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,64 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
-                    finishInitialization();
+                    CheckForImageEntriesCollection(collectionNames);
+                }).catch((err) => {
+                    throw err;
```
```suggestion
                    callback(err);
```

The `catch` below will not catch error if thrown here. `UnhandledPromiseRejectionWarning` will happen instead. Better to call the callback in this particular case.

Coteh reviewed (changes requested):
Good work! I have left some feedback for you to take a look at.

mayankbucha commented on lib/database-ops.js:
```diff
@@ -101,46 +101,64 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
+        var CheckForImageEntriesCollection = function(collectionNames) {
+            if(collectionNames.indexOf("image-entries") < 0) {
+                database.createCollection("image-entries").then((collection) => {
+                    collection.createIndex({ uploadeddate: 1 }, { 
```
Currently we are adding the index as soon as creating a new `ImageEntries` collection. So, if we create index in the else condition i.e. when the `ImageEntries` collection is already created, mongodb is throwing this warning:
`Index with name: expire_index already exists with different options`

Coteh replied:
Interesting, yeah it seems like this happens when the index is recreated with different options, e.g. if I specify a different `expireAfterSeconds` amount. Perhaps we should drop the index if it exists before we create it?

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,64 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
+                database.createCollection("image-entries").then((collection) => {
+                    collection.createIndex({ uploadeddate: 1 }, { 
+                        name: "expire_index",
+                        expireAfterSeconds: 300,
```
Perhaps we should make `expireAfterSeconds` configurable from environment variable?

mayankbucha replied:
Yes, much better. I will be setting the default to 300.

mayankbucha commented:
Hi @Coteh, I have implemented the changes. Please review

Coteh reviewed (changes requested):
Good work! Got a couple more suggestions.

Coteh commented on lib/database-ops.js:
```diff
@@ -137,9 +137,12 @@ var startDatabaseClient = function(callback, options) {
+                    createIndexForImageEntriesCollection(imageEntryCollection)
+                    finishInitialization();
+                }).catch((err) => {
+                    callback(err);
```
In the case that `dropIndex` fails due to index not existing, then we should create the index and not call the callback with error. You can check `err.codeName` and if it's "IndexNotFound" then you can call create index method and finish initialization.

Coteh commented on lib/database-ops.js:
```diff
@@ -137,9 +137,12 @@ var startDatabaseClient = function(callback, options) {
-                createIndexForImageEntriesCollection(imageEntryCollection)
-                finishInitialization();
+                imageEntryCollection.dropIndex("expire_index").then(() => {
+                    createIndexForImageEntriesCollection(imageEntryCollection)
```
```suggestion
                    createIndexForImageEntriesCollection(imageEntryCollection);
```

Coteh commented on lib/database-ops.js:
```diff
@@ -138,10 +138,16 @@ var startDatabaseClient = function(callback, options) {
                     finishInitialization();
                 }).catch((err) => {
-                    callback(err);
+                    if(err.codeName === "IndexNotFound") {
```
Would be nice to have a log statement here so we know if index is being created when there wasn't one previously.

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,79 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
-            if (err) {
-                throw err;
+        var createIndexForImageEntriesCollection = function(collection) {
+            collection.createIndex({ uploadeddate: 1 }, { 
```
Would be nice to have a log statement for when the index is created. I think it can go into the callback of `createIndex` or from its promise.

Coteh commented on lib/database-ops.js:
```diff
@@ -36,7 +36,7 @@ var insertImageIntoDatabase = function(imageEntry, callback) {
 
         imageEntry.id = shortID;
-
+        imageEntry.temp = process.env.EVALUATION_MODE ? true : false;
```
```suggestion
        imageEntry.temp = process.env.EVALUATION_MODE === "true" ? true : false;
```

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,82 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
+            else {
+                imageEntryCollection = database.collection("image-entries");
+                imageEntryCollection.dropIndex("expire_index").then(() => {
+                    createIndexForImageEntriesCollection(imageEntryCollection);
```
```suggestion
                    logger.info("expire_index for images collection dropped successfully. Creating new expire_index");
                    createIndexForImageEntriesCollection(imageEntryCollection);
```

Coteh commented on lib/database-ops.js:
```diff
@@ -101,46 +101,82 @@ var changeUserPasswordInDatabase = function(userID, hashedPassword) {
+        var checkForImageEntriesCollection = function(collectionNames) {
+            if(collectionNames.indexOf("image-entries") < 0) {
+                database.createCollection("image-entries").then((collection) => {
+                    createIndexForImageEntriesCollection(collection);
+                    imageEntryCollection = collection;
+                    logger.info("Images collection created.");
```
```suggestion
                    imageEntryCollection = collection;
                    logger.info("Images collection created.");
                    createIndexForImageEntriesCollection(collection);
```
Just reordering the log statement and setting the variable for image collection to be before creating the index. I'd be surprised if the index created log appears before the image collection created log because index creation is async, but just thought I'd move it up for sake of clarity.

mayankbucha replied:
Yes, it would be misleading :sweat_smile:.

Coteh reviewed (changes requested):
I think we're almost done! Just one more thing I noticed that I didn't before, and a couple of very small adjustments in `startDatabaseClient`.

I think `startDatabaseClient` is good enough for now. If there's any more indexes or collections that need to be created in the future, then it should probably be cleaned up to be in a chain of promises or something. I think as more things get added, we can get a good idea of how it can be abstracted.

mayankbucha commented:
I have updated the changes. I have mistakenly pushed bat file to start mongodb. I have deleted it and and committed the changes. I cannot figure out why the codecov/patch is failing.

Yes, for `startDatabaseClient` we could refactor it to be more linear when adding new indexes in future.

I am learning a lot about js, node and github by contributing to this project. Hope to continue resolving issues for the project. :smiley:

Coteh commented:
No problem. I think codecov/patch is failing because there aren't any automated tests that cover those particular lines of code. Usually, some automated unit tests or integration tests would need to be written here to cover those cases, but since there hasn't been any tests written for database-ops yet, I'll merge the changes as-is for this particular case. I have already done some manual testing for this feature and plan on doing some more testing before it reaches mater. Adding automated tests for database-ops is something I'd like to get to at some point though, as that is the preferred and the proper approach.

Coteh commented on README.md:
```diff
@@ -36,6 +36,8 @@ MONGO_INITIAL_DATABASE=<initial database for MongoDB instance>
 GA_TRACKING_ID=<Google Analytics Tracking ID (Universal Analytics)>
 LOGIN_TO_UPLOAD=true  # omit this variable if you don't want to require users to login to upload
+EVALUATION_MODE=true # omit this variable to diable automatic removal of images
+EXPIRE_AFTER_SECONDS=300 # set this to set time for the images to be stored in database in evaluation mdoe
```
```suggestion
EXPIRE_AFTER_SECONDS=300 # set this to set time for the images to be stored in database in evaluation mode (default is 300 if unspecified)
```

Coteh reviewed (changes requested):
Just one more quick change I noticed and I think we should be good to go after that.

Thanks so much for the contribution, and glad you are getting some learning out of this! :D

Coteh reviewed (approved):
LGTM! Thanks again for the contribution! :)
